### PR TITLE
[C10d] Add skeleton of LibUV backend.

### PR DESCRIPTION
**Pull request by kumpera**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105870
* #105674
* __->__ #105672
* #105184
* #105164
* #105163

This commit hooks up tcpstore creation and build flags.

### CODE REVIEW

File: torch/csrc/distributed/c10d/TCPStore.cpp [hunk 1]
```diff
@@ -236,6 +237,11 @@ TCPStore::TCPStore(std::string host, const TCPStoreOptions& opts)
     : Store{opts.timeout},
       addr_{std::move(host)},
       numWorkers_{opts.numWorkers} {
+  if (opts.useLibUV && !::c10d::detail::is_libuv_tcpstore_backend_available()) {
+    throw std::invalid_argument(
+        "use_libuv was requested but PyTorch was build without libuv support");
+  }
```
Comment (H-Huang) [inline torch/csrc/distributed/c10d/TCPStore.cpp:243]: Should probably do `TORCH_CHECK(::c10d::detail::is_libuv_tcpstore_backend_available(), "use_libuv was requested but PyTorch was build without libuv support");` instead

TORCH_CHECK goes through the pytorch exception flow and also logs the C++ stack trace

File: torch/csrc/distributed/c10d/init.cpp [hunk 2]
```diff
@@ -1301,7 +1301,8 @@ Example::
                       std::chrono::milliseconds timeout,
                       bool waitWorkers,
                       bool multiTenant,
-                      c10::optional<int> masterListenFd) {
+                      c10::optional<int> masterListenFd,
+                      bool useLibUV) {
```
Comment (H-Huang) [inline torch/csrc/distributed/c10d/init.cpp:1305]: Should useLibUV be optional for BC? What about the current usages where users are instantiating TCPStore will they need to add this argument?

Comment (kumpera) [reply torch/csrc/distributed/c10d/init.cpp:1305]: good catch.

File: torch/csrc/distributed/c10d/TCPStore.cpp (hunk 1 above)
Comment (kumpera) [reply torch/csrc/distributed/c10d/TCPStore.cpp]: fixed.

File: cmake/Dependencies.cmake [hunk 3]
```diff
@@ -1379,6 +1379,8 @@ if(USE_DISTRIBUTED AND USE_TENSORPIPE)
       set(TP_ENABLE_CUDA_IPC ON CACHE BOOL "" FORCE)
     endif()
     set(TP_BUILD_LIBUV ON CACHE BOOL "" FORCE)
+    add_compile_options(-DTORCH_USE_LIBUV)
```
Comment (fduwjj) [inline cmake/Dependencies.cmake:1382]: maybe `TORCH_DIST_USE_LIBUV`?

**Review by fduwjj — APPROVED** [review]
LGTM

**Pull request closed without merging**